sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessor.cs

```diff
@@ -681,16 +705,24 @@ public virtual async Task StopProcessingAsync(CancellationToken cancellationToke
                         }
                         try
                         {
-                            receiver = await ServiceBusSessionReceiver.CreateSessionReceiverAsync(
-                                entityPath: EntityPath,
-                                connection: _connection,
-                                options: receiverOptions,
-                                cancellationToken: cancellationToken).ConfigureAwait(false);
-
-                            await OnSessionInitializingAsync(
-                                (ServiceBusSessionReceiver)receiver,
-                                cancellationToken).
-                                ConfigureAwait(false);
+                            if (sessionId != null)
+                            {
+                                if (!sessionIdsReceiverMap.ContainsKey(sessionId))
```

Comment:
Looks like this check is still duplicated as it is in CreateAndInitializeSessionReceiver.

Also I meant for the lowest level method that calls the init event to be named CreateAndInitializeSessionReceiver. The outer one can probably GetOrCreateSessionReceiver.

Reply:
>Looks like this check is still duplicated as it is in CreateAndInitializeSessionReceiver.

Ah, you were talking about this in previous comments as well. I didn't get you are talking about adding `if (!sessionIdsReceiverMap.ContainsKey(sessionId))` check twice.

Well, this is intentional we need this double checked locking for creating the singleton receivers in a multi threaded environment.  There are two cases - 

1)  When we first check `if (!sessionIdsReceiverMap.ContainsKey(sessionId))` and session Id already exists in the map then we don't need to acquire lock for the sessionId.
2) Let's consider the case where there are 2 threads and 1 session
     * First thread comes in and checks if sessionId-1 doesn't exist in `sessionIdsReceiverMap`, it goes inside and acquire the lock.  And then starts creating a receiver.
    * While the first thread is creating a receiver, second thread comes in. sessionId-1 still doesn't exist in `sessionIdsReceiverMap` because the first thread is busy creating the receiver, so it passes the first check and waits for the lock to be released by the first thread. Whenever first thread is done creating the receiver and putting it in a `sessionIdsReceiverMap`, it releases the lock. Now, the second thread which was waiting for  the lock, acquires it and moves forward in the flow. If we don't add the second `if (!sessionIdsReceiverMap.ContainsKey(sessionId))` check, the second thread will go on creating another receiver for the same session, resulting in session lock error.

Hope this explains the logic...

For more details - https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_C#

Reply:
Yeah, I was talking about the other check in the other issue. I guess I got confused here since it is split now so it looked like a copy paste issue. It was a bit more clear that we were doing the double checking when we were doing it in the same method instead of split across two methods. But I think the comment should be enough!